### PR TITLE
[merkle tree example] Add arithmetisation module for Merkle tree inclusion proofs

### PR DESCRIPTION
**Pull request by HiddenAndBound**
### TL;DR

Added arithmetization module for Merkle tree inclusion proofs using Groestl-256 as the compression function.

### What changed?

Added a new `arithmetisation` module to the Merkle tree tests that provides a complete implementation of the constraint system for verifying Merkle tree inclusion proofs. The implementation includes:

- `MerkleTreeCS`: A constraint system with tables and channels for Merkle tree verification
- `NodesTable`: Table for handling Merkle path steps with different child pull strategies (left, right, or both)
- `RootTable`: Table for verifying root consistency
- Table filling logic for witness generation
- Boundary construction for circuit input/output consistency
- Helper functions for data conversion and flushing
- Comprehensive test suite to validate the implementation

The module uses Groestl-256 as the compression function and supports verification of multiple inclusion proofs in a single circuit.

### How to test?

The implementation includes several unit tests that can be run with:

Key tests include:
- `check_ordering` and `check_state_orientation`: Verify correct byte ordering
- `test_nodes_table_constructor` and `test_root_table_constructor`: Verify table construction
- `test_node_table_filling` and `test_root_table_filling`: Test table filling logic
- `test_merkle_tree_cs_fill_tables`: Test the complete constraint system table filling
- `test_merkle_tree_cs_end_to_end`: End-to-end test with random Merkle trees

### Why make this change?

This implementation provides a reference for implementing Merkle tree inclusion proofs in the Binius M3 framework. It demonstrates how to:
1. Define constraint systems for cryptographic primitives
2. Handle complex data structures like Merkle trees
3. Implement witness generation for zero-knowledge proofs
4. Create proper boundary conditions for circuit inputs/outputs

The module serves both as a test for the Merkle tree implementation and as a reference for developers implementing similar cryptographic primitives in the framework.

### CODE REVIEW

**Comment by HiddenAndBound** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/IrreducibleOSS/binius/508?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#508** <a href="https://app.graphite.dev/github/pr/IrreducibleOSS/binius/508?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/IrreducibleOSS/binius/508?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#507** <a href="https://app.graphite.dev/github/pr/IrreducibleOSS/binius/507?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#506** <a href="https://app.graphite.dev/github/pr/IrreducibleOSS/binius/506?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#505** <a href="https://app.graphite.dev/github/pr/IrreducibleOSS/binius/505?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#504** <a href="https://app.graphite.dev/github/pr/IrreducibleOSS/binius/504?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>merge-ready</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/IrreducibleOSS?ref=merge-queue-instructions-in-stack-comment&prId=508).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by pepyakin — COMMENTED** [review]
Overall, this looks good but a lot of cosmetic nits.

Cosmetic nits are more important here IMO as examples are essentially the face of the product.

File: crates/m3/tests/merkle_tree.rs [hunk 1]
```diff
@@ -490,3 +490,778 @@ mod model {
...
+						),
+					);
+					table.pull(
+						nodes_channel,
+						to_node_flush(
+							root_id_upcasted,
+							right_packed,
+							child_depth_upcasted,
+							right_index_upcasted,
+						),
```
[the remark this answers is not in the document]
Comment (HiddenAndBound) [reply crates/m3/tests/merkle_tree.rs:864]: Good point
